subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AbstractClassGenerator.java

```diff
@@ -905,10 +915,32 @@ void applyTo(ClassGenerationVisitor visitor) {
                     visitor.applyReadOnlyManagedStateToGetter(property, getter.method);
                 }
             }
+            for (PropertyMetadata property : readOnlyNestedProperties) {
+                visitor.applyManagedStateToProperty(property);
+                for (MethodMetadata getter : property.getters) {
+                    visitor.applyNestedManagedStateToGetter(property, getter.method);
+                }
+            }
             if (!hasFields) {
-                visitor.addManagedMethods(mutableProperties, readOnlyProperties);
+                visitor.addManagedMethods(
+                    mutableProperties,
+                    concat(readOnlyProperties, readOnlyNestedProperties)
+                );
             }
         }
+
+        private static List<PropertyMetadata> concat(List<PropertyMetadata> first, List<PropertyMetadata> second) {
+            if (first.isEmpty()) {
+                return second;
+            }
+            if (second.isEmpty()) {
+                return first;
+            }
+            return ImmutableList.<PropertyMetadata>builderWithExpectedSize(first.size() + second.size())
+                    .addAll(first)
+                    .addAll(second)
+                    .build();
+        }
```

Comment:
I looked into using `Iterable`, though we use the `size` of the collection somewhere downstream. Since we generate a lot of classes, I was too afraid to change the code here to break a performance test.

Reply:
I am using `Iterables.concat` now. It wasn't hard after all.